### PR TITLE
LPD-26672 Only glob properties that represent a url path

### PR DESCRIPTION
**Pull request by thiago-buarqque**
https://liferay.atlassian.net/browse/LPD-26672

# Issue
If a developer adds a `*` as part of the value of a frontend client extension property that is not supposed to be treated as a path (e.g `/path/**/*.js`), the current implementation of "expandWildcards" will break. 

# Approach
Verify if the property **name includes "url"** for frontend client extensions **and** if the property value **is not a URL**.

The common thing is that all frontend properties that support glob contain 'url' in their name, so we could make that a pattern and rely on that.

The logic for the Validator is copied from the portal kernel Validator class, since we can't just use that here.

### Discarded approach
We could add a list of properties that support glob, but there are properties that can either be a URL or a glob path.

cc: @ethib137 @drewbrokke

### CODE REVIEW

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:sf

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/Validator.java
Comment (algobob) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/Validator.java:1]: What it validates? "Validator" is too generic. If its verifying String nullity/empty, there's a StringUtils.isBlank in Workspace project for this.

Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/Validator.java:1]: That's actually the goal, we have a Validator in the kernel, it's used to validate a variety of things. But I agree with you, it's not well written, will fix it.

[the remark this answers is not in the document]
Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/Validator.java:1]: Ops, my bad

Comment (algobob) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/Validator.java:1]: Cool. It seems there are multiple Utils in workspace. Maybe we can use isBlank for check the string and create a HTTPUtil for the isUrl part? :thinking:

[the remark this answers is not in the document]
Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/Validator.java:1]: Done

Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/Validator.java:1]: Done

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/HTTPUtil.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+				new URL(url);
+
+				return true;
+			}
+			catch (MalformedURLException malformedURLException) {
+			}
+		}
+
+		return false;
+	}
```
Comment (drewbrokke) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/HTTPUtil.java:42]: I think we can just put this into StringUtil.
Also, I don't think there's ever a case in this context where we'd want to `acceptRootRelative`, right? If not then we can remove the overload and remove the short-circuit.

**Review by drewbrokke — CHANGES_REQUESTED** [review]

**Comment by drewbrokke** [comment]
ci:test:gradle-plugins

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/HTTPUtil.java (hunk 1 above)
Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/HTTPUtil.java:42]: Done, waiting current CI run to finish to make sure nothing in the logic is wrong

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:gradle-plugins

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
CI is passing, please @drewbrokke let me know if you have any considerations. Thanks

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/StringUtil.java [hunk 2]
```diff
@@ -91,6 +96,24 @@ public static boolean isBlank(String s) {
 		return false;
 	}
 
+	public static boolean isUrl(String url) {
+		if (!isBlank(url)) {
```
Comment (drewbrokke) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/StringUtil.java:100]: Can you flatten the nesting here by using an early return for `if isBlank`?

**Review by drewbrokke — CHANGES_REQUESTED** [review]
@thiago-buarqque this looks good, just one more change then I think we're good to go.

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/StringUtil.java (hunk 2 above)
Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/StringUtil.java:100]: Sure, will do. Thanks! :)

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:gradle-plugins

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
@drewbrokke all tests passed, should we forward this?

**Comment by drewbrokke** [comment]
ci:test:workspaces

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/StringUtil.java [hunk 3]
```diff
@@ -5,9 +5,14 @@
 
 package com.liferay.gradle.plugins.workspace.internal.util;
 
+import com.liferay.petra.string.CharPool;
```
Comment (drewbrokke) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/StringUtil.java:8]: Sorry I missed this before, @thiago-buarqque, but I'd like to keep all `petra` dependencies out of workspace for now. You can add the colon as a constant field directly here in `StringUtil` (there are a few others like it).

**Review by drewbrokke — CHANGES_REQUESTED** [review]
Sorry @thiago-buarqque, one more thing I missed earlier.

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/StringUtil.java (hunk 3 above)
Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/util/StringUtil.java]: @drewbrokke Done

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:workspaces

**Comment by thiago-buarqque** [comment]
ci:test:gradle-plugins

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:workspaces

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/task/CreateClientExtensionConfigTask.java [hunk 4]
```diff
@@ -401,11 +401,17 @@ private void _expandWildcards(Map<String, Object> typeSettings) {
...
-			if ((currentValue instanceof String) &&
-				_isWildcardValue((String)currentValue)) {
+			if (currentValue instanceof String) {
+				String currentValueString = (String)currentValue;
 
-				entry.setValue(
-					_getMatchingPaths(staticDirPath, (String)currentValue));
+				String key = StringUtil.toLowerCase(entry.getKey());
+
+				if (key.contains("url") &&
```
Comment (drewbrokke) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/task/CreateClientExtensionConfigTask.java:409]: @thiago-buarqque do we not also want to check whether the key contains "url" down below in the block starting with `currentValue instanceof List`? Or is that not a valid use-case?

Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/task/CreateClientExtensionConfigTask.java:409]: Not a valid use case, as of today we don't have a list of urls

Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/task/CreateClientExtensionConfigTask.java:409]: Ops, we do. Missed it, my bad

Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/task/CreateClientExtensionConfigTask.java:409]: @drewbrokke fixed it, sorry about that

**Comment by thiago-buarqque** [comment]
ci:test:gradle-plugins

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:stable

**Comment by thiago-buarqque** [comment]
@drewbrokke can I forward this?

**Review by drewbrokke — APPROVED** [review]

**Comment by drewbrokke** [comment]
Looks good @thiago-buarqque, thanks!

**Comment by thiago-buarqque** [comment]
ci:forward:force

**Pull request closed without merging**